### PR TITLE
Read/List with new table data when dbNewtable flag is turned on

### PR DESCRIPTION
**Pull request by mikeyang923**
Read/List with new table data when dbNewtable flag is turned on

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java [hunk 1]
```diff
@@ -19,6 +19,7 @@ public class MySqlNamedBlobDbConfig {
   private static final String PREFIX = "mysql.named.blob.";
   public static final String DB_INFO = PREFIX + "db.info";
   public static final String DB_TRANSITION = PREFIX + "db.transition";
+  public static final String DB_NEW_TABLE = PREFIX + "db.newtable";
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:22]: can you rename this to something more descriptive, something like `db.read.from.new.table`.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 2]
```diff
@@ -256,9 +257,9 @@ public CompletableFuture<NamedBlobRecord> get(String accountName, String contain
         new GetTransactionStateTracker(remoteDatacenters, localDatacenter);
     return executeTransactionAsync(accountName, containerName, true, (accountId, containerId, connection) -> {
       long startTime = System.currentTimeMillis();
-      NamedBlobRecord record, recordV2;
+      NamedBlobRecord record, recordV2 = null;
       record = run_get(accountName, containerName, blobName, option, accountId, containerId, connection);
-      if (config.dbTransition) {
+      if (config.dbTransition || config.dbNewtable) {
```
Comment (justinlin-linkedin) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:262]: There are two problems with this code.
1. even when reading from new table, we are still throw exception for old table db operation and catching exception for new table operations.
2. reading from new table is a concept that is valid only when we are in the middle of transition. So it's not necessary to put these two boolean statements together.

I would recommend doing something like this
```
interface GetRecord {
    NamedBlobRecord do(String accountName, String containerName, String blobName, GetOption option, short accountId, short containerId, Connection connection) throws Exception;
}

NamedBlobRecord primary, secondary;
GetRecord primaryGetter = this.run_get, secondaryGetter = this.run_get_v2;
if (config.dbNewTable) {
    primaryGetter = this.run_get_v2;
    secondaryGetter = this.run_get;
}

primary = primaryGetter.do(xxxx);
if (config.dbTransition) {
    // do secondary
}
return primary;
```

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java (hunk 1 above)
Comment (mikeyang923) [reply ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java]: Done.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java (hunk 2 above)
Comment (mikeyang923) [reply ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java]: Those two problems are addressed in 7826a9b.
Thank you for the interface suggestion. If we go with that interface way, then we need to create multiples interfaces (for Get, List, Put, Delete). And I prefer to code the logic like below to be more explicit:

```
      try {
        result = operation_on_old_table(...);
      } catch (Exception eOld) {
        if (!config.dbNewtable) {
          throw eOld;
        }
      }

      if (config.dbTransition) {
        try {
          resultV2 = operation_on_new_table(...);
          ....
        } catch (Exception e) {
          if (config.dbNewtable) {
            throw e;
          }
          ....
        }
      }
     ...
      return config.dbNewtable? resultV2: result;
```

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java [hunk 3]
```diff
@@ -55,14 +56,22 @@ public class MySqlNamedBlobDbConfig {
    * A flag on whether to turn on DB Transition run
    */
   @Config(DB_TRANSITION)
-  @Default("false")
+  @Default("true")
```
Comment (SophieGuo410) [inline ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:59]: Why do we change the default value? Have you tested the prod and corp env already?

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java [hunk 4]
```diff
@@ -55,14 +56,22 @@ public class MySqlNamedBlobDbConfig {
...
+  @Config(DB_RELY_ON_NEW_TABLE)
+  @Default("false")
+  public final boolean dbRelyOnNewtable;
+
   public MySqlNamedBlobDbConfig(VerifiableProperties verifiableProperties) {
     this.dbInfo = verifiableProperties.getString(DB_INFO);
     this.localPoolSize = verifiableProperties.getIntInRange(LOCAL_POOL_SIZE, 5, 1, Integer.MAX_VALUE);
     this.remotePoolSize = verifiableProperties.getIntInRange(REMOTE_POOL_SIZE, 1, 1, Integer.MAX_VALUE);
     this.listMaxResults = verifiableProperties.getIntInRange(LIST_MAX_RESULTS, 100, 1, Integer.MAX_VALUE);
     this.dbTransition = verifiableProperties.getBoolean(DB_TRANSITION, false);
```
Comment (SophieGuo410) [inline ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:74]: This default should keep same as what you defined above.

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java [hunk 5]
```diff
@@ -55,14 +56,22 @@ public class MySqlNamedBlobDbConfig {
...
-  @Default("false")
+  @Default("true")
   public final boolean dbTransition;
 
+  /**
+   * A flag on whether to use new table data
+   */
+  @Config(DB_RELY_ON_NEW_TABLE)
+  @Default("false")
+  public final boolean dbRelyOnNewtable;
```
Comment (SophieGuo410) [inline ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:67]: nit: dbRelyOnNew**T**able

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java (hunk 3 above)
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:59]: +1

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java (hunk 5 above)
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:67]: +1

File: ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java (hunk 4 above)
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/config/MySqlNamedBlobDbConfig.java:74]: +1

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 6]
```diff
@@ -256,16 +257,27 @@ public CompletableFuture<NamedBlobRecord> get(String accountName, String contain
...
     return executeTransactionAsync(accountName, containerName, true, (accountId, containerId, connection) -> {
       long startTime = System.currentTimeMillis();
-      NamedBlobRecord record, recordV2;
-      record = run_get(accountName, containerName, blobName, option, accountId, containerId, connection);
+      NamedBlobRecord record = null, recordV2 = null;
+      try {
+        record = run_get(accountName, containerName, blobName, option, accountId, containerId, connection);
+      } catch (Exception eOld) {
+        if (!config.dbRelyOnNewtable) {
+          throw eOld;
```
Comment (ankagrawal) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:265]: Maybe add a metric here?

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 7]
```diff
@@ -256,16 +257,27 @@ public CompletableFuture<NamedBlobRecord> get(String accountName, String contain
...
         try {
           recordV2 = run_get_v2(accountName, containerName, blobName, option, accountId, containerId, connection);
-          if (!record.equals(recordV2)) {
+          if (record != null && !recordV2.equals(record)) {
             logger.warn("NAMED GET: Data Inconsistence: old record='{}', new record='{}'", record, recordV2);
             metricsRecoder.namedDataInconsistentGetCount.inc();
           }
         } catch (Exception e) {
+          if (config.dbRelyOnNewtable) {
+            throw e;
```
Comment (ankagrawal) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:279]: Metric here too? We can use this to surface if any requests are failing due to new or old value.

File: ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java [hunk 8]
```diff
@@ -287,60 +299,91 @@ public CompletableFuture<Page<NamedBlobRecord>> list(String accountName, String
...
       if (config.dbTransition) {
         try {
           deleteResultV2 = run_delete_v2(accountName, containerName, blobName, accountId, containerId, connection);
-          if (!deleteResult.equals(deleteResultV2)) {
+          if (deleteResult != null && !deleteResultV2.equals(deleteResult)) {
             logger.warn("NAMED DELETE: Data Inconsistence: old record='{}', new record='{}'", deleteResult, deleteResultV2);
             metricsRecoder.namedDataInconsistentDeleteCount.inc();
           }
         } catch (Exception e) {
+          if (config.dbRelyOnNewtable) {
```
Comment (ankagrawal) [inline ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java:384]: No tests for these cases?

**Review by SophieGuo410 — APPROVED** [review]

**Pull request merged**